src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -429,6 +433,15 @@ - (void)_handleReportEnd
 {
     os_unfair_lock_lock(&self->_lock);
     _estimatedStartTimeFromGeneralDiagnosticsUpTime = nil;
+// For unit testing only
+#ifdef DEBUG
+    id delegate = _weakDelegate.strongObject;
+    if (delegate && [delegate respondsToSelector:@selector(unitTestReportEndForDevice:)]) {
```

Comment:
I would be much happier if we did the respondsToSelector check inside the async block, because calling out into arbitrary code with our lock held seems scary.

I know this is debug-only, etc, but still.

Reply:
This code is doing:
1. A pre-flight check that the delegate responds to the unit test callback
2. Schedule a block on a different queue.

And so the arbitrary unit test code won't be run while holding the MTRDevice lock.

This is actually really clean and not problematic.

Reply:
I see what you mean. I've made the change.